presto-memory-context/src/main/java/com/facebook/presto/memory/context/AggregatedMemoryContext.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory.context;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.OptionalLong;
+
+import static com.facebook.presto.memory.context.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class AggregatedMemoryContext
+{
+    @Nullable
+    final AggregatedMemoryContext parentMemoryContext;
+    @Nullable
+    private final MemoryReservationHandler reservationHandler;
+    @GuardedBy("this")
+    long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+    private final OptionalLong guaranteedMemory;
+
+    public AggregatedMemoryContext()
+    {
+        this(null, OptionalLong.empty());
+    }
+
+    public AggregatedMemoryContext(MemoryReservationHandler reservationHandler, OptionalLong guaranteedMemoryInBytes)
+    {
+        this(null, reservationHandler, guaranteedMemoryInBytes);
+    }
+
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext, MemoryReservationHandler reservationHandler, OptionalLong guaranteedMemory)
+    {
+        checkArgument(reservationHandler == null || parentMemoryContext == null, "at least one of reservationHandler and parentMemoryContext must be null");
+        this.parentMemoryContext = parentMemoryContext;
+        this.reservationHandler = reservationHandler;
+        this.guaranteedMemory = requireNonNull(guaranteedMemory, "guaranteedMemory is null");
+    }
+
+    public AggregatedMemoryContext newAggregatedMemoryContext()
+    {
+        return new AggregatedMemoryContext(this, null, OptionalLong.empty());
+    }
+
+    public LocalMemoryContext newLocalMemoryContext()
+    {
+        return new SimpleLocalMemoryContext(this);
+    }
+
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    synchronized ListenableFuture<?> updateBytes(long bytes)
+    {
+        ListenableFuture<?> future = NOT_BLOCKED;
+        // delegate to parent if exists
+        if (parentMemoryContext != null) {
```

Comment:
I can only repeat: please separate this two different logics into two classes, converting `AggregatedMemoryContext`  to an interface.

Reply:
I am hesitating to make this change, because I don't think splitting this into two classes will buy us much. It may help splitting the two logic as you point out, but the complexity of this class hierarchy and the complexity of using these classes will increase.

Reply:
I see this as following. You can make `AggregatedMemoryContext` an abstract class with:
- `private long usedBytes`, package-visible `directUpdateBytes` (you need this method so that `SimpleLocalMemoryContext` doesn't need to synchronize on instances of this class)
- methods like `getBytes()`, `newLocalMemoryContext()`
- package-visible constructor
- two package-visible subclasses implementing update logic
- public factory methods here, in this class (`newRootContext`, .. or something like that)

This won't change usage patterns (callers won't notice any change except for construction, which will be cleaner). I think this would be a change for better.